aeternity/ae_mdw
# feat: account presence based on aex9 balance

## What

Once transfer event is something manually emitted from the contract code, account presence might be updated after any transfer as a best effort to keep track of 'account-to-contract' mapping.

## Why

Fixes #186 as a best effort as the complete mapping (of all accounts with balance in the contract) depends on contract code to emit the event.

An account might appear in:
/aex9/balances/:contract_id
but not in:
/aex9/balances/account/:account_id

## Validation steps

1. With `master` branch, this account is initially not mapped to any contract (before running the migration, which runs automatically):
http://localhost:4000/aex9/balances/account/ak_2CuCtmoExsuW7iG5BbRAoncSZAoZt8Abu21qrPUhqfFNin75iL
http://localhost:4000/aex9/balances/account/ak_eYHyfKSZiU3nzDTJBkrXAzVszoxDU4DHDw5xz6f2i5YVME234
2. Run the Mdw with `186-index-account-presence` branch. Wait for migration to get all aex9 contracts balances and update the accounts presence (this might take about 10 minutes).
3. Repeat the requests of step 1 and the mapped accounts balance is shown:
![image](https://user-images.githubusercontent.com/44991200/135424522-1619d5e7-c1ed-49ef-b4a2-8514c8975da5.png)
![image](https://user-images.githubusercontent.com/44991200/135424448-6fafbe63-6127-4ad5-952a-4da820b0a4fc.png)

Expected migration indexed count is (for Mdw snapshot of August 24th 2021):
Indexed 14 records in 504s

## Additional Notes

Please notice the actual transfer calls are not shown in the request as there is no txi (mapping not after a transfer event with the account).

These aex9 contracts are out of the scope of this PR as a call to get their store/balance returns "Out of gas" error:

Out of gas for ct_2M4mVQCDVxu6mvUrEue1xMafLsoA1bgsfC3uT95F3r1xysaCvE
Out of gas for ct_21oSi6uMMcnnyxc1KTP8MReBPVvBQQ9oeoA7aexQBsF2o16HwT
Out of gas for ct_2czjMpbeeds9Geb272UfiEvhQHc8QZM5AM8jngwpqdMnWNDP3h
Out of gas for ct_25JPViJPDPjroiL1wr1ngWJ8pK5ykN7x3LqGMWPCuC6PtLjzKT
Out of gas for ct_2mzKUzT55wz4wYUmH4s3dHf3YBprDLi92uxvEah8fe3BaLfkBU
Out of gas for ct_2bFV4kxtmUeKTF5Eb5STfnAx6UgxzBqTL1pYLF1GBhcmoQAhLf
Out of gas for ct_BmCPDh6TK3KtWdwKEJrSCnrvghbkjekuTTh7B5tLGduF6rTFG
Out of gas for ct_227nuD7YiHL7pJ63FP1CnbhKgU9ueWKP77sA26huE47s31tuqQ
Out of gas for ct_jyw2fRpr6srNwSqE6FJFyZSD1upD91eZ5ZZKZfBhuDmBxTU53
Out of gas for ct_VxFemyM2ML15ApWrFqBKrarVbNfBASPBs1SC1zMuwaWrnrTSy
Out of gas for ct_2GPhKGbRxZaaJtqRTnxLk5UDSNNEtRt3tyWuKVQeoVrerRfhHa
Out of gas for ct_7UfopTwsRuLGFEcsScbYgQ6YnySXuyMxQWhw6fjycnzS5Nyzq
Out of gas for ct_M9yohHgcLjhpp1Z8SaA1UTmRMQzR4FWjJHajGga8KBoZTEPwC
Out of gas for ct_BwJcRRa7jTAvkpzc2D16tJzHMGCJurtJMUBtyyfGi2QjPuMVv
Out of gas for ct_uGk1rkSdccPKXLzS259vdrJGTWAY9sfgVYspv6QYomxvWZWBM
Out of gas for ct_26GAggwrHJLw45vopxu1HA51t6pw4xjzqvq66kh1WPtZjNXMyL
